marlowe-playground-client/src/Halogen/ActusBlockly.purs

```diff
@@ -0,0 +1,172 @@
+  = Resize a
+  | SetError String a
+
+data ContractFlavour = FS | F
+
+data BlocklyAction
+  = Inject String (Array BlockDefinition)
+  | GetTerms ContractFlavour
+
+data BlocklyMessage
+  = Initialized
+  | CurrentTerms ContractFlavour String
+
+type Slots
+  = ()
+
+type DSL m a
+  = HalogenM BlocklyState BlocklyAction Slots BlocklyMessage m a
+
+blockly :: forall m. MonadEffect m => String -> Array BlockDefinition -> Component HTML BlocklyQuery Unit BlocklyMessage m
+blockly rootBlockName blockDefinitions =
+  mkComponent
+    { initialState: const { actusBlocklyState: Nothing, generator: Nothing, errorMessage: Nothing }
+    , render
+    , eval:
+      H.mkEval
+        { handleQuery
+        , handleAction
+        , initialize: Just $ Inject rootBlockName blockDefinitions
+        , finalize: Nothing
+        , receive: const Nothing
+        }
+    }
+
+handleQuery :: forall m a. MonadEffect m => BlocklyQuery a -> DSL m (Maybe a)
+handleQuery (Resize next) = do
+  mState <- use _actusBlocklyState
+  case mState of
+    Just state ->
+      pure
+        $ ST.run
+            ( do
+                workspaceRef <- STRef.new state.workspace
+                Blockly.resize state.blockly workspaceRef
+            )
+    Nothing -> pure unit
+  pure $ Just next
+
+handleQuery (SetError err next) = do
+  assign _errorMessage $ Just err
+  pure $ Just next
+
+handleAction :: forall m. MonadEffect m => BlocklyAction -> DSL m Unit
+handleAction (Inject rootBlockName blockDefinitions) = do
+  blocklyState <- liftEffect $ Blockly.createBlocklyInstance rootBlockName (ElementId "actusBlocklyWorkspace") (ElementId "actusBlocklyToolbox")
+  let
+    _ =
+      ST.run
+        ( do
+            blocklyRef <- STRef.new blocklyState.blockly
+            workspaceRef <- STRef.new blocklyState.workspace
+            Blockly.addBlockTypes blocklyRef blockDefinitions
+            Blockly.initializeWorkspace blocklyState.blockly workspaceRef
+        )
+
+    generator = buildGenerator blocklyState
+  modify_ _ { actusBlocklyState = Just blocklyState, generator = Just generator }
+
+handleAction (GetTerms flavour) = do
+  res <-
+    runExceptT do
+      blocklyState <- ExceptT <<< map (note $ unexpected "BlocklyState not set") $ use _actusBlocklyState
+      generator <- ExceptT <<< map (note $ unexpected "Generator not set") $ use _generator
+      let
+        workspace = blocklyState.workspace
+        rootBlockName = blocklyState.rootBlockName
+      block <- except <<< (note $ unexpected ("Can't find root block" <> rootBlockName)) $ getBlockById workspace rootBlockName
+      except <<< lmap (\x -> "This workspace cannot be converted to code: " <> (show x)) $ blockToCode block generator
+  case res of
+    Left e -> assign _errorMessage $ Just e
+    Right contract -> do
+      assign _errorMessage Nothing
+      raise $ CurrentTerms flavour $ contract
+  where
+  unexpected s = "An unexpected error has occurred, please raise a support issue: " <> s
```

Comment:
I mentioned this to Kris, can we give detail of how, maybe "raise an issue at https://github.com/input-output-hk/plutus/issues" or something

Reply:
Kris had some ideas about this code, I find it a bit messy:
"
Hmm…I’d start by packaging up `ExceptT <<< map (note` as a helper…
I don’t think the let-bindings are pulling their weight.
"